Add game_theory.md

CY-chang opened the pull request:
[#55](https://github.com/NTHU-CP/NTHU-CPP/issues/55)

- [x] Read [Contribute to NTHU-CPP](https://nthu-cp.github.io/NTHU-CPP/others/contribution.html).
- [x] Rebase to the latest main branch.
- [ ] List all the references.
- [x] Successfully build the website by mdBook with no error after the change.
- [x] Exclude any irrelevant files.
- [x] Link to the issue which is related to your change.
- [x] Review the content by myself.
- [ ] Pass tests/CI.

harry900831 commented on src/math/game_theory.md:
```diff
@@ -0,0 +1,296 @@
+
+而一個有關賽局的題目，我們會想知道在某個狀態，當前玩家能不能採用一個最佳化的策略，使得對手不論如何應對，都無法阻止該玩家獲勝。這種狀態稱為 **winning state**。反之，若玩家在某狀態，不論採取何種行動，都無法阻止對手以某個策略取勝的話，就稱為**losing state**。換句話說，winning state至少存在一種行動，使得遊戲局勢落入losing state。而losing state不論採取什麼行動，都會使局勢落入winning state。
+
+### Example
```
I believe providing an easier example to introduce game theory is better.
For example:

兩人 100顆石頭，輪流每人每次可以取 1 ~ 7 顆，沒石頭拿的就輸了，請問誰穩贏。

harry900831 replied:
我認為這題不需要真的有OJ 連結，這只是讓讀者迅速理解這章節大概是什麼樣子的內容與讓他們腦力激盪而已。

harry900831 commented on src/math/game_theory.md:
```diff
@@ -0,0 +1,296 @@
+
+而在題目範圍太大或是想不到解法時，打表找規律也是一個不錯的辦法。
+
+## Reference
```
references

harry900831 reviewed (changes requested):
我大致瀏覽過了，我先將其轉乘 draft PR，等你修改好並附上所有練習題後，再將PR 設定成 ready for review.
1. 英文字左右需要空白，除非是標點符號之類得特殊字元。
2. 目前內容有點少，我相信應該有更多定理或是技巧能講。
3. 這個主題的重點會是範例題目，請盡量涵蓋各種與賽局相關的題目與應用。

CY-chang commented:
@harry900831 
Hi, I've added some exercises and their editorials. For the introduction part, I couldn't find an Online Judge with the problem you mentioned. So I replaced it with a similar one on CSES. Should I add more exercise problems, or should I focus on expanding the content related to the Nim and Grundy numbers section (not quite sure what to add) ?
Thanks for your feedback!

harry900831 reviewed (changes requested):
I think you can rename this Topic to "Nim and Sprague-Grundy Theorem". If you don't know what to add for the theorem part, I think you can add more example problem. 

There's a lot of cool and interesting problem that is related to game theory and it doesn't require any special theorem.

For example: 

一個 n*m 的巧克力在桌上，兩人輪流拿一片起來橫斷或直斷，直到沒有 1*1 以外的塊，沒得操作的人輸

or

ZeroJudge] b902. 肉墊遊戲


(Although, I believe there should be some theorem behind those problem)

harry900831 commented:
If you have no idea, read more online resources may be helpful.

harry900831 commented:
The content looks great at first glance, I will do detail review in a few days.

harry900831 commented on src/math/nim_and_sg_value.md:
```diff
@@ -0,0 +1,550 @@
+
+### Rules
+
+有 \\( N \\) 堆石頭，每堆有 \\( a_i \\) 個。玩家 A，B 輪流取石頭，每次可以選擇一堆，並從該堆取出任意數量的石頭(不能不取)。沒有石頭可以取的玩家落敗。
```
A、B

harry900831 commented on src/math/nim_and_sg_value.md:
```diff
@@ -0,0 +1,550 @@
+
+### Strategy
+
+將每堆的石頭個數進行 \\( bitwise\ XOR \\) 運算，得到的答案可以拿來判斷局勢。
+
+得到的答案若為 \\( 0 \\)，代表當前的局勢為 losing state，答案不為 \\( 0 \\) 則為 winning state。
```
將每堆的石頭個數進行 xor 若得到的數值為 0 代表是 losing state 先手必輸，或不為 0 則為 winning state，先手必贏。

harry900831 commented on src/math/nim_and_sg_value.md:
```diff
@@ -0,0 +1,550 @@
+
+這邊提出一個數學歸納法的證明。為了方便後續說明，將 \\( XOR \\) 得出的結果命名為 \\( s \\)。
+
+#### Base Case
+
+若沒有任何石頭，則 \\( s\\) 為 \\( 0 \\)，玩家無法採取行動，為一個 losing state。
+
+#### Inductive Case
+
+若當前為 losing state，亦即 \\( s=0 \\)。因為遊戲規定每回合都必須取石頭。若從有 \\( x \\) 塊的石頭堆取石頭，設取完剩 \\( y \\) 塊 \\( (x>y) \\)，那麼取完後的 \\( XOR \\) 答案為 \\( (s \oplus x \oplus y) = (x \oplus y)\neq 0 \\)，一定會使局勢變為 winning state。
+
+若當前為 winning state，即 \\( s \neq 0 \\)。考慮 \\( s \\) 的 most significant bit (假設從右到左為第 \\( k \\) 個 bit)，那麼至少存在某一堆石頭的個數(設為 \\( x \\) )，且該 \\( x \\) 從右到左的第 \\( k \\) 個 bit 也同樣為 \\( 1 \\) (否則 \\( s \\) 的該 bit 就不會是 \\( 1 \\) )。那麼從該堆石頭中取石頭，取到剩下 \\( (s \oplus x ) \\) 個，那麼取完之後 \\( XOR \\) 的值就會變成 \\( (s \oplus x \oplus (s \oplus x)) = 0 \\)，是一個 losing state。而該堆取完後剩 \\( (s \oplus x) \\) 個，該數的第 \\( k \\) 個 bi ...
+
+綜上所述以及數學歸納法，且遊戲在有限的回合數內會結束，因此該策略是正確的。
```
看不出這數學歸納法是在歸納什麼

harry900831 commented on src/math/nim_and_sg_value.md:
```diff
@@ -0,0 +1,550 @@
+>
+> 有 \\( N \\) 堆棍子，每堆有 \\( x_i \\) 根棍子。玩家 A, B 輪流採取行動。每次行動可以選擇其中一堆棍子，並從中拿走任意正整數根棍子。沒有棍子可以取的一方落敗。問若從 A 開始，且雙方都採取最佳化策略，那麼誰將獲勝。
+
+- \\( 1 \leq N \leq 2*10^5 \\)
```
\times

harry900831 commented on src/math/nim_and_sg_value.md:
```diff
@@ -0,0 +1,550 @@
+
+### Example
+
+> [CSES 1099 - Stair Game](https://cses.fi/problemset/task/1099)
+>
+> 有 \\( N \\) 階的階梯，編號為第 \\( 1 \sim N \\) 階，每階上都放有非負整數顆球。玩家 \\( A\, B \\) 輪流採取行動，由 \\( A \\) 開始。每回合可以選擇一個整數 \\( x \\) \\( (x > 1) \\)，並從第 \\( x \\) 階取任意正整數顆球並移到第 \\( x - 1 \\) 階。到最後，沒有任何方法移動球的一方落敗。題目則問若雙方都採取最佳化策略時，誰會獲勝。
```
missing problem size

harry900831 commented on src/math/nim_and_sg_value.md:
```diff
@@ -0,0 +1,550 @@
+
+經由打表和一些觀察可以發現，若將**偶數階階梯**的球數進行 \\( XOR \\) 運算，若為 \\( 0 \\) 代表該 state 為 losing state，若不為 \\( 0 \\) 則為 winning state。
+
+該如何驗證這個策略的正確性呢?這邊提供一個與 Nim Game 有關的想法。將偶數階階梯的球數視為 Nim Game 每堆的石頭數，且每回合的行動是可以任取一堆並增加該堆石頭數的。而對於一個偶數 \\( K\ (K \geq 2) \\)，若當時第 \\( K + 1 \\) 階有 \\( x \\) 顆球，則代表第 \\( K \\) 階的石頭堆在當時的回合最多可以新增 \\( x \\) 塊石頭。而因為能新增的石頭數量是**有限**的，因此可以採用同上個章節"Nim Game Extension"的策略。
```
解釋不夠清楚

harry900831 commented on src/math/nim_and_sg_value.md:
```diff
@@ -0,0 +1,550 @@
+
+## Sprague-Grundy Theorem
+
+在介紹此定理以前，先引入一個符號 \\( MEX \\)。對於一個集合 \\( s \\)，\\( MEX \lbrace s \rbrace \\) 的定義為不屬於 \\(s \\) 的數字中最小的非負整數。例如 \\( MEX \lbrace 0, 1, 3, 5 \rbrace = 2 \\)，\\( MEX \lbrace 1, 2, 3 \rbrace = 0 \\)。根據 Sprague-Grundy Theorem，一個不偏賽局可以被轉換成 Nim Game 的形式。對於一個狀態(局勢)，我們定義它的 **Grundy Number** (SG value)為 \\( MEX \lbrace g_1, g_2, ..., g_n \rbrace \\)，這裡的 \\( g_i \\) 代表當前的 state 可以轉移到某 \\( n \\) 個不同的 state，且 \\( ...
```
需要清楚的 state 出這個 theorem 是什麼

可能可以用 ::: :::

CY-chang replied:
不太了解。是要將內文出現的"Sprague-Grundy Theorem"都改格式嗎，還是說把這段以正式的 theorem 來呈現

harry900831 replied:
這段已正式的 theorem 來呈現

:::
Sprague-Grundy Theorem

(It's definition)
:::

harry900831 commented on src/math/nim_and_sg_value.md:
```diff
@@ -0,0 +1,550 @@
+
+## Sprague-Grundy Theorem
+
+在介紹此定理以前，先引入一個符號 \\( MEX \\)。對於一個集合 \\( s \\)，\\( MEX \lbrace s \rbrace \\) 的定義為不屬於 \\(s \\) 的數字中最小的非負整數。例如 \\( MEX \lbrace 0, 1, 3, 5 \rbrace = 2 \\)，\\( MEX \lbrace 1, 2, 3 \rbrace = 0 \\)。根據 Sprague-Grundy Theorem，一個不偏賽局可以被轉換成 Nim Game 的形式。對於一個狀態(局勢)，我們定義它的 **Grundy Number** (SG value)為 \\( MEX \lbrace g_1, g_2, ..., g_n \rbrace \\)，這裡的 \\( g_i \\) 代表當前的 state 可以轉移到某 \\( n \\) 個不同的 state，且 \\( ...
```
那 非 0 的grundy number 怎麼定義？

harry900831 replied:
喔喔 我看到下面的例題就看懂了

harry900831 commented on src/math/nim_and_sg_value.md:
```diff
@@ -0,0 +1,550 @@
+
+另外，若遊戲方式為好幾個子遊戲同時進行。我們可以採用與 Nim Game 相似的作法，將若干個子遊戲的 Grundy Number 使用 \\( XOR \\) 來進行合併，得出整個遊戲的狀態。
+
+定理的證明可以利用數學歸納法與 \\( XOR \\) 的性質。但這裡提供一個直觀的想法供讀者參考:
```
Get 不到這個直觀的想法

harry900831 commented on src/math/nim_and_sg_value.md:
```diff
@@ -0,0 +1,550 @@
+
+在介紹此定理以前，先引入一個符號 \\( MEX \\)。對於一個集合 \\( s \\)，\\( MEX \lbrace s \rbrace \\) 的定義為不屬於 \\(s \\) 的數字中最小的非負整數。例如 \\( MEX \lbrace 0, 1, 3, 5 \rbrace = 2 \\)，\\( MEX \lbrace 1, 2, 3 \rbrace = 0 \\)。根據 Sprague-Grundy Theorem，一個不偏賽局可以被轉換成 Nim Game 的形式。對於一個狀態(局勢)，我們定義它的 **Grundy Number** (SG value)為 \\( MEX \lbrace g_1, g_2, ..., g_n \rbrace \\)，這裡的 \\( g_i \\) 代表當前的 state 可以轉移到某 \\( n \\) 個不同的 state，且 \\( ...
+
+另外，若遊戲方式為好幾個子遊戲同時進行。我們可以採用與 Nim Game 相似的作法，將若干個子遊戲的 Grundy Number 使用 \\( XOR \\) 來進行合併，得出整個遊戲的狀態。
```
這句話放在這裡我相信讀者也會看不懂

harry900831 replied:
這感覺是需要證明的東西？

harry900831 commented on src/math/nim_and_sg_value.md:
```diff
@@ -0,0 +1,550 @@
+
+除了可以用 \\( XOR \\) 合併多個子遊戲的 Grundy Number 以外，若一個行動將遊戲分裂成若干個子遊戲，那麼同樣也可以利用 \\( XOR \\) 將分裂的子遊戲合併，並以 \\( MEX \\) 函數得到該狀態的 Grundy Number。光看文字敘述可能不好理解，我們直接藉由 Grundy's Game 這個例子來說明。
+
+> 一開始有 \\( x \\) 根棍子放在同一堆。玩家 A, B 輪流採取行動。每次行動可以選擇其中一堆並將其分成兩堆，而且這兩堆的大小不能相同。沒有合法的行動者落敗。問雙方都採取最佳化策略，那麼誰將獲勝。
```
有online judge 嗎？

harry900831 commented on src/math/nim_and_sg_value.md:
```diff
@@ -0,0 +1,550 @@
+
+\\( f(10) = MEX \lbrace f(1) \oplus f(9), f(2) \oplus f(8), f(3) \oplus f(7) , f(4) \oplus f(6) \rbrace \\)。
+
+以 \\( x = 5 \\) 為例，因為 \\( f(5) > 0 \\)，代表當前的狀態是 winning state。同時也代表著玩家可以採取某個行動，使遊戲狀態的 Grundy Number 變為 \\( 0 \\)，也就是 losing state。在這個局勢，玩家只要把這堆棍子分成 \\( 1 \\) 根與 \\( 4 \\) 根，則 Grundy Number 會變成 \\( f(1) \oplus f(4) = 0 \\)。
```
summary for this problem?

harry900831 commented on src/math/nim_and_sg_value.md:
```diff
@@ -0,0 +1,550 @@
+>
+> 給一棵含有 \\( N \\)個節點的樹。玩家 A, B 輪流採取行動。每次可以選擇一條邊，並把整棵樹切成兩個部分。不含有編號 \\( 1 \\) 節點的聯通塊會被移除。若某方無法採取合法的行動，則落敗。問若從 A 開始，且兩人都採取最佳化策略，那麼誰將獲勝。
+
+在一次行動後，不含有 \\( 1 \\) 號節點的聯通塊會被移除。因此把根定在 \\( 1 \\) 號節點似乎是個合理的想法。接著要思考的是如何定義狀態。樹具有可以遞迴的特性。如果我們有子樹的狀態，有沒有辦法把那些答案合併，並求出父節點的答案呢? 基於上述的推論，我們可能會聯想到 Grundy Number。若一個子樹的根結點編號為 \\( x \\)，我們定義在這棵子樹上玩相同的遊戲時，Grundy Number 為 \\( f_x \\)。那麼對於一個父節點 \\( u \\)，對於他所有的子節點 \\( v_i \\)，如果我們知道所有 \\( f_{v_i} \\) 的值，那麼我們能否求出 \\( f_u \\)。
```
what's the difference between grundy number and sg value?

CY-chang replied:
Two terms mean the same thing. Should I choose one and use it throughout the whole article?

harry900831 commented on src/math/nim_and_sg_value.md:
```diff
@@ -0,0 +1,550 @@
+>
+> 給一棵含有 \\( N \\)個節點的樹。玩家 A, B 輪流採取行動。每次可以選擇一條邊，並把整棵樹切成兩個部分。不含有編號 \\( 1 \\) 節點的聯通塊會被移除。若某方無法採取合法的行動，則落敗。問若從 A 開始，且兩人都採取最佳化策略，那麼誰將獲勝。
+
+在一次行動後，不含有 \\( 1 \\) 號節點的聯通塊會被移除。因此把根定在 \\( 1 \\) 號節點似乎是個合理的想法。接著要思考的是如何定義狀態。樹具有可以遞迴的特性。如果我們有子樹的狀態，有沒有辦法把那些答案合併，並求出父節點的答案呢? 基於上述的推論，我們可能會聯想到 Grundy Number。若一個子樹的根結點編號為 \\( x \\)，我們定義在這棵子樹上玩相同的遊戲時，Grundy Number 為 \\( f_x \\)。那麼對於一個父節點 \\( u \\)，對於他所有的子節點 \\( v_i \\)，如果我們知道所有 \\( f_{v_i} \\) 的值，那麼我們能否求出 \\( f_u \\)。
```
f_x or f(x)? 
Please be consistent throughout whole article

harry900831 commented on src/math/nim_and_sg_value.md:
```diff
@@ -0,0 +1,550 @@
+
+- \\( 1 \leq N, L_i, R_i \leq 100 \\)
+
+<details><summary>Solution</summary>
```
missing space before and after Solution

harry900831 commented on src/math/nim_and_sg_value.md:
```diff
@@ -0,0 +1,550 @@
+
+</details>
+
+<details><summary> Solution Code </summary>
```
solution code should be inside solution

harry900831 commented on src/math/nim_and_sg_value.md:
```diff
@@ -0,0 +1,550 @@
+
+<details><summary>Solution</summary>
+
+第一眼看到題目可能會覺得狀態不太容易定義。但再想想會發現若把遊戲狀態定義成目前可選擇的範圍是 \\( x \sim y \\)。那麼取完一個區間後，會得到另外兩個可以選的小區間，也就是兩個同類型的子遊戲。這與前面 Grundy Number 章節的概念頗為相似。考慮將狀態定義為 \\( dp\[x\]\[y\] \\)，代表當現在可選的範圍是 \\( x \sim y \\) 時，狀態的 Grundy Number 會是多少。那麼採用 top-down DP 的方式來填表，就能得到整個遊戲的 SG-value，也就能得到最後的答案。
```
dp[x][[y] -> f(x, y)

harry900831 commented on src/math/nim_and_sg_value.md:
```diff
@@ -0,0 +1,550 @@
+
+將一堆石頭分成若干堆後，會分成若干個獨立且同類的遊戲。於是我們可以聯想到 Grundy Number。首先，我們必須知道狀態會如何轉移。\\( x \\) 塊石頭會被分成數量有 \\( y 塊, y + 1 塊, y + 2 塊…… \\) 的石頭堆。因為總和還是 \\( x \\)，將這 \\( x \\) 塊分出來的堆數最多只會是 \\( O(sqrt(x)) \\) 的量級。因此可以枚舉會被分成幾堆，並 \\( O(1) \\) 求出 \\( y \\)。這樣狀態轉移只需要 \\( O(sqrt(N)) \\) 的時間。而總共有 \\( N \\) 個狀態，總時間複雜度為 \\( O(N * sqrt(N)) \\)。
+
+另外，我們必須求出轉移到的狀態的 SG value，也就是被分出來的那幾堆的 SG value 做 \\( XOR \\)運算。但慢慢做需要 \\( O(sqrt(N)) \\)，總複雜度變成 \\( O(N *sqrt(N)* sqrt(N)) = O(N ^ 2) \\)，會超時。因此我們需要一個更快的作法。觀察到被拆出來的石頭堆會形成公差為 \\( 1 \\) 的等差數列。因此我們可以構造出 Grundy Number 的前綴和，並 \\( O(1) \\) 計算出 \\( f_y \oplus f_{y+1} \oplus f_{y+2} …… \\)。在花 \\( O(sqrt(N) * 1) \\) 計算出能轉移到的狀態的 SG value 後，再花 \\( O(sqrt(N)) \\) 的時間求 \\( MEX \\) 函數的答案，就能求出現在這個狀態的 SG value  ...
```
render error

harry900831 commented on src/math/game_theory.md:
```diff
@@ -0,0 +1,428 @@
+
+\\( n = 1 \\) 與 \\( n = 2 \\) 這兩個 corner case 因為情況較特殊，以下證明會撇除這兩種情況，後續再另外討論。
+
+在遊戲剛開始時，雙方差距的格子數為 \\( 2n - 2 \\)。若 A 每回合只前進一步，就算 B 每回合以最少的步數，也就是前進兩步，拖延他撞上 A 的時間。每回合雙方的距離仍會減一。在第 \\( 2n - 4 \\) 個回合結束後，雙方的差距格子數會減少至 \\( 2 \\)。於是在第 \\( 2n - 3 \\) 回合時 B 就會碰到 A，使遊戲結束。而陷阱的部分因為 A 總共只移動了 \\( 2n - 4 \\) 步，沒有走到 B 的起點，因此不用擔心會踩到陷阱。
+
+這邊簡單說明一下為什麼 A 不會在選擇一個策略後，中途突然改變策略。雖然聽起來有點理所當然，但筆者認為仍需要提一下，使後續的證明更加順暢且完整。若 A 一開始選擇大步數追擊，而後續突然改變為上述守株待兔的策略，那麼兩者的距離差距比一開始來得更遠，因此該策略一定比上述的直接小步數等待還來得差。反之，若 A 一開始不論是因為甚麼原因，像是躲避陷阱，或是誘敵等，選擇小步數追擊，而之後突然改變策略變為大步數追擊。該策略也不會直接從一開始就直接選擇大步數追擊來得佳。原因為就算 A 有可能因為改變步數而成功躲避一些陷阱，但遇到陷阱的代價也只是該回合少移動一步，那麼直接以最大步數進行每一回合，一定不會比改變步數來得差。
+
+接著分析 A 採取追擊的策略。首先 A 在不會踩到陷阱的前提下每回合都會前進 \\( 4 \\) 步。而 B 在看到 A 大步數前進，根據上方的說明，B 於是知道 A 採取的是追擊策略，因此除了第一步以外，B 的每個回合都會前進 \\( 3 \\) 步。因此要討論的只有 B 以 \\( 2 \\) 步開局或是以 \\( 3 \\) 步開局這兩種情況了。若 B 以 \\( 3 \\) 步開局，代表每個回合，A 與 B 的距離只會被拉近 \\( 1 \\)，而且因為每回合是由 B 先動，因此至少需要 \\( 2n - 2 \\) 個回合，A 才能獲勝，比守株待兔的策略來得差。而若 B 以 \\( 2 \\) 步開局，第一個回合結束後，雙方距離差 \\( 2n - 4 \\)。若雙方能在每回合拉近 \\( 1 \\) 格的距離，那麼 A 同樣也能在恰好第 \\( 2n - 3 \\) 回合結束 ...
```
例如這裡每一段的文字都很多，閱讀起來會不太順。
可以考慮搭配圖並必精簡你的文字同時不犧牲傳遞的資訊量。

harry900831 reviewed (changes requested):
- use \times instead of *
- In my opinion, there's a lot of part that can be shorter. Please try to use less word but with clear meaning to explain stuff.
- You should have detail explanation on the theorem part.
- It's very great that you have a lot of problem as an example
- there're a lot of parts that didn't render successfully, please check it.

CY-chang commented on src/SUMMARY.md:
```diff
@@ -24,6 +24,8 @@
     - [數論分塊](math/sqrt_decomposition.md)
     - [杜教篩](math/du_sieve.md)
     - [Arithmetic Function Revisit](math/revisit_arithmetic_function.md)
+  - [Game Theory](math/game_theory.md)
+  - [Nim and Sprague-Grundy Theorem](math/nim_and_sg_value.md)
```
Under misc seems better. I'll move the location of the files.

harry900831 commented:
Are you ready for review?

CY-chang commented:
> Are you ready for review?

Yes, most of the problems are fixed.

fffelix-huang commented on src/misc/game_theory.md:
```diff
@@ -0,0 +1,462 @@
+
+- 不變量(Invariant): 一種不會改變的性質。
+
+在賽局的題目中，必勝方通常可以維持某種不變量。觀察並定義出不變量往往是解賽局題目的關鍵。以 Introduction 拿石頭的題目為例，不變量是在兩人操作結束後，剩餘數量維持在 \\( 8 \\) 的倍數。而後面章節會介紹的 Nim Game，不變量則為兩人操作結束後，所有堆數的 \\( XOR \\) 等於 \\( 0 \\)。
```
讀者還不知道 Nim Game 是甚麼，堆數是什麼。或許這邊可以先不用提到 Nim Game，到 Nim Game 的 chapter 再分析他的 invariant？

harry900831 replied:
我覺得這邊需要至少兩個例子
如果找不到其他比 nim game 更好的例子 就維持就好
反正他有提是後面章節

fffelix-huang commented on src/misc/game_theory.md:
```diff
@@ -0,0 +1,462 @@
+
+剩下的只有兩疊皆為偶數的情況了。若當前玩家採取第一種行動，會取出奇數個墊子，使局勢變為一奇一偶，也就是 winning state。但這麼做無疑是一個錯誤的策略，因此雙方玩家只會採取第二種行動。於是在雙方的策略都被確定的情況，我們可以開始探討給定兩疊的數量，誰會獲勝。若兩疊的數量為 \\( x, y\\)，假設 \\( x \geq y \\)，那麼在之後的 \\( \frac{x}{y} \\) 回合都會從數量為 \\( x \\) 那疊中取出墊子。而數量則會變為 \\( x\ mod\ y \\) 與 \\( y \\)。接著再從數量為 \\( y \\) 的那疊開始取，以此類推。可以發現過程和求最大公因數時使用的輾轉相除法極為相似，因此若暴力將行動模擬完，複雜度也只會是 \\( log(max\lbrace x, y \rbrace) \\)。而取到最後，會有其中一疊數量歸零，若另 ...
+
+時間複雜度為: \\( O(log(max\lbrace x, y \rbrace)) \\)
```
\log

fffelix-huang commented on src/misc/nim_and_sg_value.md:
```diff
@@ -0,0 +1,564 @@
+
+前面章節有提到，將每堆石頭進行 \\( XOR \\) 運算得到的答案，可以拿來判斷局勢。這種性質同樣可以應用在 Grundy Number 上。將若干個子遊戲的 Grundy Number 使用 \\( XOR \\) 來進行合併，可以得出整個遊戲的 Grundy Number。同樣地，若為 \\( 0 \\) 代表是 losing state。若不為 \\( 0 \\) 則為 winning state。
+
+詳細的證明可以參考[這篇文章](https://zhuanlan.zhihu.com/p/20611132)。但這裡提供一個直觀的想法供讀者參考: 試著把不偏賽局的 Grundy Number 與 Nim Game 做連結。Nim Game 每回合的行動可以從數量為 \\( x(x>0) \\) 的石頭堆取出任意數量，在取後的剩餘石頭數可以是 \\( 0 \sim x-1 \\) 的任意一個。同樣地，一個 Grundy Number 為 x 的 state，對於任意一個 \\( i\ (x-1\ge i\ge0, \ i \in Z) \\)，都至少存在一種轉移方法，使玩家在一回合內轉移到 SG value 為 \\( i \\) 的 state。
```
SG value 或是 Grundy Number，選一個用

fffelix-huang replied:
x 沒 render 到

fffelix-huang reviewed (changes requested):
都看過了，再檢查看看有沒有地方沒有 render 到~

harry900831 commented:
@CY-chang Hi, your PR is almost ready to be merged, have you fixed all the comments?

CY-chang commented:
> @CY-chang Hi, your PR is almost ready to be merged, have you fixed all the comments?

Yes, I believe all of them are fixed.
